core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala

```diff
@@ -0,0 +1,356 @@
+   * @return true if the block did not already exist, false otherwise. If this returns false, then
+   *         no new locks are acquired. If this returns true, a write lock on the new block will
+   *         be held.
+   */
+  def lockNewBlockForWriting(
+      blockId: BlockId,
+      newBlockInfo: BlockInfo): Boolean = synchronized {
+    logTrace(s"Task $currentTaskAttemptId trying to put $blockId")
+    val actualInfo = infos.getOrElseUpdate(blockId, newBlockInfo)
+    if (actualInfo eq newBlockInfo) {
+      actualInfo.writerTask = currentTaskAttemptId
+      writeLocksByTask.addBinding(currentTaskAttemptId, blockId)
+      true
+    } else {
+      false
+    }
+  }
+
+  /**
+   * Release all lock held by the given task, clearing that task's pin bookkeeping
+   * structures and updating the global pin counts. This method should be called at the
+   * end of a task (either by a task completion handler or in `TaskRunner.run()`).
+   *
+   * @return the ids of blocks whose pins were released
+   */
+  def unlockAllLocksForTask(taskAttemptId: TaskAttemptId): Seq[BlockId] = {
+    val blocksWithReleasedLocks = mutable.ArrayBuffer[BlockId]()
+    synchronized {
+      writeLocksByTask.remove(taskAttemptId).foreach { locks =>
+        for (blockId <- locks) {
+          infos.get(blockId).foreach { info =>
+            assert(info.writerTask == taskAttemptId)
+            info.writerTask = -1
+          }
+          blocksWithReleasedLocks += blockId
+        }
+      }
+      notifyAll()
+    }
+    val readLocks = readLocksByTask.get(taskAttemptId)
+    readLocksByTask.invalidate(taskAttemptId)
+    readLocks.entrySet().iterator().asScala.foreach { entry =>
+      val blockId = entry.getElement
+      val lockCount = entry.getCount
+      blocksWithReleasedLocks += blockId
+      synchronized {
+        get(blockId).foreach { info =>
+          info.readerCount -= lockCount
+          assert(info.readerCount >= 0)
+        }
+      }
+    }
+    synchronized {
+      notifyAll()
+    }
+    blocksWithReleasedLocks
+  }
+
+  /**
+   * Returns the number of blocks tracked.
+   */
+  def size: Int = synchronized {
+    infos.size
+  }
+
+  /**
+   * Return the number of map entries in this pin counter's internal data structures.
+   * This is used in unit tests in order to detect memory leaks.
+   */
+  private[storage] def getNumberOfMapEntries: Long = synchronized {
+    size +
+      readLocksByTask.size() +
+      readLocksByTask.asMap().asScala.map(_._2.size()).sum +
+      writeLocksByTask.size +
+      writeLocksByTask.map(_._2.size).sum
+  }
+
+  /**
+   * Returns an iterator over a snapshot of all blocks' metadata. Note that the individual entries
+   * is this iterator are mutable and thus may reflect blocks that are deleted while the iterator
+   * is being traversed.
+   */
+  def entries: Iterator[(BlockId, BlockInfo)] = synchronized {
+    infos.iterator.toArray.toIterator
```

Comment:
This doesn't match the comment: `this iterator are mutable and thus may reflect blocks that are deleted while the iterator is being traversed.` since they are copied to an array